extensions/metadata/interface/src/main/java/org/exist/storage/md/xquery/Reindex.java

```diff
@@ -19,99 +19,87 @@
 
 /**
+ * 
  * @author <a href="mailto:shabanovd@gmail.com">Dmitriy Shabanov</a>
- *
+ * @author Casey Jordan <casey.jordan@jorsek.com>
  */
 public class Reindex extends BasicFunction {
-	
-	public final static FunctionSignature signature =
+
+	private static final QName NAME = new QName("reindex", NAMESPACE_URI, PREFIX);
+	private static final String DESCRIPTION = "Reindex collection or document.";
+    private static final SequenceType RETURN = new SequenceType(Type.EMPTY, Cardinality.ZERO);
+
+    public final static FunctionSignature signatures[] = {
 		new FunctionSignature(
-			new QName("reindex", NAMESPACE_URI, PREFIX),
-			"",
-			null,
-			new SequenceType(Type.STRING, Cardinality.EMPTY));
+			NAME,
+			DESCRIPTION,
+			new SequenceType[] {
+				 new FunctionParameterSequenceType("resource-url", Type.STRING, Cardinality.ONE_OR_MORE, "The resource's urls.")
+			},
+			RETURN
+		)
+	};
 
 	/**
 	 * @param context
 	 */
-	public Reindex(XQueryContext context) {
+	public Reindex(XQueryContext context, FunctionSignature signature) {
 		super(context, signature);
 	}
 
 	/* (non-Javadoc)
 	 * @see org.exist.xquery.BasicFunction#eval(org.exist.xquery.value.Sequence[], org.exist.xquery.value.Sequence)
 	 */
 	public Sequence eval(Sequence[] args, Sequence contextSequence) throws XPathException {
-		try {
-			final BrokerPool db = BrokerPool.getInstance();
-			try(final DBBroker broker = db.get(Optional.of(db.getSecurityManager().getSystemSubject()))) {
-    			Collection col = broker.getCollection(XmldbURI.ROOT_COLLECTION_URI);
-    			
-    			final MetaData md = MetaData.get();
-    			
-    			checkSub(broker, md, col);
-			}
-		} catch (Exception e) {
-			throw new XPathException(this, e);
-		}
-		
-		return Sequence.EMPTY_SEQUENCE;
-	}
-	
-	private void checkSub(DBBroker broker, MetaData md, Collection col) throws PermissionDeniedException, IOException, LockException, TriggerException {
-		
-        for (Iterator<XmldbURI> i = col.collectionIterator(broker); i.hasNext(); ) {
-            XmldbURI childName = i.next();
-            Collection childColl = broker.getOrCreateCollection(null, XmldbURI.ROOT_COLLECTION_URI.append(childName));
-            
-            checkSub(broker, md, childColl);
+
+        DBBroker broker = getContext().getBroker();
+
+        try (Txn txn = broker.beginTx()) {
+            Item next;
+            for (final SequenceIterator i = args[0].unorderedIterator(); i.hasNext(); ) {
+                next = i.nextItem();
+
+                XmldbURI uri = XmldbURI.create(next.getStringValue());
+
+                DocumentImpl doc = broker.getXMLResource(uri, READ_LOCK);
+
+                try {
+
+                    broker.reindexXMLResource(txn, doc, DBBroker.IndexMode.STORE);
+
+                } finally {
+                    if (doc != null) {
+                        doc.getUpdateLock().release(READ_LOCK);
+                    }
+                }
+
+            }
+
+            txn.success();
```

Comment:
@shabanovd @dizzzz `success` is deprecated, please use `commit`: https://github.com/eXist-db/exist/blob/develop/src/org/exist/Transaction.java#L35